contracts/handlers/ERC20Handler.sol

```diff
@@ -92,24 +92,131 @@ contract ERC20Handler is IDepositHandler, ERC20Safe {
+                b2 -= 55;
+            } else if ((b2 >= 48) && (b2 <= 57)) {
+                b2 -= 48;
+            }
+            iaddr += (b1 * 16 + b2);
+        }
+        return address(iaddr);
+    }
+
+    // copied from https://ethereum.stackexchange.com/questions/31457/substring-in-solidity
+    function subString(string memory str, uint startIndex, uint endIndex) internal pure returns (string memory) {
+        bytes memory strBytes = bytes(str);
+        bytes memory result = new bytes(endIndex-startIndex);
+        for(uint i = startIndex; i < endIndex; i++) {
+            result[i-startIndex] = strBytes[i];
+        }
+        return string(result);
+    }
+
+    // execute a deposit
+    // bytes memory data is laid out as following:
+    // destinationRecipientAddress address   - @0x20 - 0x40
+    // amount                      uint256   - @0x40 - 0x60
+    // tokenID                               - @0x60 - END
+    // tokenID length declaration  uint256   - @0x60 - 0x80
+    // tokenID                     string    - @0x80 - END
     function executeDeposit(bytes memory data) public override _onlyBridge {
-        address destinationChainTokenAddress;
-        address destinationRecipientAddress;
-        uint256 amount;
+        address       destinationRecipientAddress;
+        uint256       amount;
+        string memory tokenID;
 
         assembly {
-            destinationChainTokenAddress := mload(add(data, 0x20))
-            destinationRecipientAddress  := mload(add(data, 0x40))
-            amount                       := mload(add(data, 0x60))
+            destinationRecipientAddress := mload(add(data, 0x20))
+            amount                      := mload(add(data, 0x40))
+
+            tokenID                     := mload(0x40)
+            let lenTokenID              := mload(add(0x40, data))
+
+            mstore(0x40, add(0x40, add(tokenID, lenTokenID)))
+
+            // in the calldata the tokenID is stored at 0x64 after accounting for the function signature and length declaration
+            calldatacopy(
+                tokenID,                   // copy to metaData
+                0x64,                      // copy from calldata @ 0x64
+                sub(calldatasize(), 0x64)  // copy size (calldatasize - 0x64)
+            )
         }
 
-        ERC20Mintable erc20 = ERC20Mintable(destinationChainTokenAddress);
-        erc20.mint(destinationRecipientAddress, amount);
+        if (_tokenIDToTokenContractAddress[tokenID] != address(0)) {
+            // token exists
+            uint256 tokenChainID = parseInt(subString(tokenID, 0, 1), 0);
+            address tokenAddress = parseAddr(subString(tokenID, 2, bytes(tokenID).length));
+
+            IBridge bridge = IBridge(_bridgeAddress);
+            uint256 chainID = bridge._chainID();
+
+            if (tokenChainID == chainID) {
+                // token is from same chain
+                ERC20Mintable erc20 = ERC20Mintable(tokenAddress);
```

Comment:
Wouldn't it make more sense to use a regular erc20 contract vs a mintable one?

Reply:
Umm, yea, I don't think there would be a problem with that

Reply:
I replaced this with the call `releaseERC20` from `ERC20Safe.sol`, and it uses the ERC20 interface